collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java

```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
+    private boolean insertRow(TreeElement element, TreeElement instanceIDElement, File instanceFile, String sheetTitle) throws Exception {
+        List<Object> columnTitles = getColumnTitles(element, instanceIDElement);
+        if (!isNumberOfColumnsValid(columnTitles.size())) {
             return false;
         }
 
-        // make sure column names are legal
-        for (String n : columnNames) {
-            if (!isValidGoogleSheetsString(n)) {
-                outcome.results.put(id,
-                        Collect.getInstance().getString(R.string.google_sheets_invalid_column_form,
-                                n));
+        List<List<Object>> sheetCells = getSheetCells(sheetTitle);
+        if (sheetCells != null && !sheetCells.isEmpty()) { // we are editing an existed sheet
+            if (isAnyColumnHeaderEmpty(sheetCells.get(0))) {
+                // Insert a header row again to fill empty headers
+                sheetsHelper.updateRow(spreadsheetId, sheetTitle + "!A1",
+                        new ValueRange().setValues(Collections.singletonList(columnTitles)));
+                sheetCells = getSheetCells(sheetTitle); // read sheet cells again to update
+            }
+            if (isAnyMissingColumn(sheetCells.get(0), columnTitles)) {
+                outcome.results.put(id, Collect.getInstance().getString(R.string.google_sheets_missing_columns));
                 return false;
             }
+        } else { // new sheet
+            resizeSheet(getSheetId(sheetTitle), columnTitles.size());
+            insertRow(columnTitles, sheetTitle);
+            sheetCells = getSheetCells(sheetTitle); // read sheet cells again to update
         }
 
-        // parses the instance file and populates the answers and photos
-        // hashmaps.
-        HashMap<String, String> answersToUpload = new HashMap<>();
-        HashMap<String, String> mediaToUpload = new HashMap<>();
-
-        try {
-            processInstanceXML(instanceFile, answersToUpload, mediaToUpload);
-        } catch (FormException e) {
-            outcome.results.put(id,
-                    Collect.getInstance().getString(R.string.google_repeat_error));
-            return false;
-        } catch (XmlPullParserException | IOException e) {
-            Timber.e(e, "Exception thrown while parsing the file");
-            outcome.results.put(id, e.getMessage());
-            return false;
+        HashMap<String, String> answers = getAnswers(getChildElements(element), instanceFile);
+        if (shouldRowBeInserted(answers)) {
+            if (element.isRepeatable()) {
+                answers.put(getElementTitle(instanceIDElement), instanceIDElement.getValue().getDisplayText());
+            }
+            insertRow(prepareListOfValues(sheetCells.get(0), columnTitles, answers), sheetTitle);
         }
+        return true;
+    }
 
+    private void insertRow(List<Object> rowElements, String sheetName) throws IOException {
```

Comment:
I don't think this method is helpful. If a method name is overloaded, it should be to have different variants with different parameter lists. Here, the two `insertRow` methods are not related in this way. Making the helper call and catching the exception in the primary `insertRow` would be clearer.